src/python/pants/backend/tools/semgrep/rules.py

```diff
@@ -0,0 +1,230 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Iterable
+
+from pants.backend.python.util_rules import pex
+from pants.backend.python.util_rules.pex import PexRequest, VenvPex, VenvPexProcess
+from pants.core.goals.lint import LintResult, LintTargetsRequest
+from pants.core.util_rules.partitions import Partition, Partitions
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests, PathGlobs, Snapshot
+from pants.engine.process import FallibleProcessResult, ProcessCacheScope
+from pants.engine.rules import Get, MultiGet, Rule, collect_rules, rule
+from pants.engine.target import AllTargets, Target, Targets
+from pants.engine.unions import UnionRule
+from pants.option.global_options import GlobalOptions
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize, softwrap
+
+from .subsystem import SemgrepFieldSet, SemgrepSubsystem
+from .target_types import SemgrepRuleSourceField
+
+logger = logging.getLogger(__name__)
+
+
+class SemgrepLintRequest(LintTargetsRequest):
+    field_set_type = SemgrepFieldSet
+    tool_subsystem = SemgrepSubsystem
+
+
+@dataclass(frozen=True)
+class PartitionMetadata:
+    config_files: frozenset[SemgrepRuleSourceField]
+    ignore_files: Snapshot
+
+    @property
+    def description(self) -> str:
+        return ", ".join(sorted(field.file_path for field in self.config_files))
+
+
+_IGNORE_FILE_NAME = ".semgrepignore"
+
+
+def warn_about_ignore_files_if_required(ignore_files: Snapshot, semgrep: SemgrepSubsystem) -> None:
+    non_root_files = sorted(name for name in ignore_files.files if name != _IGNORE_FILE_NAME)
+    if non_root_files and not semgrep.acknowledge_nested_semgrepignore_files_are_not_used:
+        # https://github.com/returntocorp/semgrep/issues/5669
+        logger.warning(
+            softwrap(
+                f"""
+                Semgrep does not obey {_IGNORE_FILE_NAME} outside the working directory, which is
+                the build root when run by pants. These files may not have the desired effect:
+                {', '.join(non_root_files)}
+
+                Set `acknowledge_nested_semgrepignore_files_are_not_used = true` in the `[semgrep]`
+                section of pants.toml to silence this warning.
+                """
```

Comment:
- How likely is this to happen? 
- And can we not partition by config AND ignore?
- Does the tool itself not warn? It seems weird to hand-hold the user a bit here. We don't do that for other tools, like `black`.

Reply:
1. I don't think it's particularly likely, however, I feel it's worth putting some effort into:
   1. It's the only way to stop semgrep running on particular files, given there's no per-target `skip_semgrep=True` flag (and I don't know how to add one, other than adding it to all `Target`s!).
   2. I could imagine the first instinct of someone trying to ignore things in a particular subfolder might be to try a `.gitignore` approach and add `.semgrepignore` there.
   3. If someone is already running semgrep in subfolders, they may have nested `.semgrepignore`s (e.g. if merging a polyrepo into a monorepo, that each have `.semgrepignore`s)
2. I don't think we can _easily_ partition by ignore:
   1. It'd require synthesising a single `.semgrepignore` managing the hierarchy and location appropriately.
   2. It can be retrofitted in future, though, and this flag ensures that we won't silently break someone: if someone has nested ignores, and set this flag, and then we add nested handling in future, we can check the flag to maintain the no-nesting behaviour (deprecated).
3. No, I don't think it warns (and even if it did, that might be hidden until https://github.com/pantsbuild/pants/issues/18618#issuecomment-1487825598). I see your point though, and 1a above is what has pushed me to make this a pants concern.

> How bad would it be to PR this without the semgrepignore and then add it in in a separate PR?

Yeah, I'll do this.

Reply:
(Separated, and stashed in https://github.com/huonw/pants/pull/1 so I don't forget about it.)